### How to run this notebook

* Online with [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/joaquimrcarvalho/cipf-comtrade.git/main?labpath=comtrade-api.ipynb):
https://mybinder.org/v2/gh/joaquimrcarvalho/cipf-comtrade.git/main?labpath=comtrade-api.ipynb
* In your own machine: check https://github.com/joaquimrcarvalho/cipf-comtrade.git


# Acesso à API UN Comtrade

> The UN COMTRADE is the largest depository of international trade data. It contains well over 5 billion data records since 1962 and is available publicly on the internet. In addition, it offers public and premium data API for easier integration/download.

Informação geral: https://unstats.un.org/wiki/display/comtrade/What+is+UN+Comtrade

API Site: https://comtradedeveloper.un.org 
API specs: https://comtradedeveloper.un.org/api-details#api=comtrade-v1&operation=get-get


Manual metodológico 
* alterações a partir de 2019: https://comtrade.un.org/data/MethodologyGuideforComtradePlus.pdf
* conceitos: https://unstats.un.org/unsd/trade/eg-imts/IMTS%202010%20(English).pdf

Codebook:
* https://unstats.un.org/wiki/display/comtrade/New+Comtrade+FAQ+for+First+Time+Users?preview=/125141443/135004360/ComtradePlus%20-%20data%20items%20-%2017%20Mar%202020.xlsx
* Documentação geral de códigos e categorias: https://unctadstat.unctad.org/EN/Classifications.html

Python package (low level calls):
* https://github.com/uncomtrade/comtradeapicall

Ver também:
* https://wits.worldbank.org/wits/wits/witshelp/Content/Data_Retrieval/T/Intro/B1.Working_with_Trade_Data.htm
* https://wits.worldbank.org/wits/wits/witshelp/Content/Data_Retrieval/T/Intro/B2.Imports_Exports_and_Mirror.htm

## Setup comtrade tools

In [2]:
import comtradetools as comtrade

comtrade.setup()

## Get a free API Key

To access the UN Comtrade via API, you need an access key,
otherwise the results are limited to 500 lines.

To obtain the access key:
* Register at https://comtradedeveloper.un.org/
* Go to _Products_ 
* Select "Premium Individual APIs" (https://comtradedeveloper.un.org/product#product=dataapis)
* Choose _Subscribe to "comtrade - v1"_  
* Wait for the email with the API key (takes a few days)
* Copy the key to the location indicated in the `config.ini` file before running the rest of the notebook.
  running the rest of the notebook.

---

Para aceder à UN Comtrade via API sem limites é necessário uma chave de acesso,
de outro modo os resultados são limitados a 500 linhas.

Para obter a chave de acsso:
* Registo em https://comtradedeveloper.un.org/
* Ir para _Products_ 
* Selecionar "Premium Individual APIs" (https://comtradedeveloper.un.org/product#product=dataapis)
* Escolher _Subscribe to "comtrade - v1"_  
* Esperar pelo email com a chave da API key (demora alguns dias
* Copia a chave para o local indicado no ficheiro `config.ini` antes
  de executar o resto do notebook.


In [2]:
import comtradetools as comtrade

print('Current config file with APIKEY:',comtrade.CONFIG_FILE)

Current config file with APIKEY: config.ini


## init comtrade tools with API Key

In [3]:
import comtradetools as comtrade

APIKEY = comtrade.get_api_key()

comtrade.init(APIKEY, force_init=False)

INFO:root:Loading HS codes from support/harmonized-system.csv


## Columns in the dataset

The following table describes the columns in the dataset.

In [4]:
import comtradetools as comtrade

print(comtrade.DATA_ITEM_DF.to_markdown())

|    |   Unnamed: 0 | dataItem                 | description                                                                              | isInDataAPICommodity   | isInDataAPIServices   | isInBulkFileFinal   | isInBulkFileTariffline   |
|---:|-------------:|:-------------------------|:-----------------------------------------------------------------------------------------|:-----------------------|:----------------------|:--------------------|:-------------------------|
|  0 |            0 | datasetCode              | Combination of several keys to identify specific datasetCode                             | True                   | True                  | True                | True                     |
|  1 |            1 | typeCode                 | Product type: Goods or Services                                                          | True                   | True                  | True                | True                     |
|  2 |            2 | freqCode                 |

### List of columns in the dataset

In [5]:
import comtradetools as comtrade

cols_all = comtrade.DATA_ITEM_DF.dataItem.unique()
cols_all


array(['datasetCode', 'typeCode', 'freqCode', 'refPeriodId', 'refYear',
       'refMonth', 'period', 'reporterCode', 'reporterISO',
       'reporterDesc', 'flowCategory', 'flowCode', 'flowDesc',
       'partnerCode', 'partnerISO', 'partnerDesc', 'partner2Code',
       'partner2ISO', 'partner2Desc', 'classificationSearchCode',
       'classificationCode', 'isOriginalClassification', 'cmdCode',
       'cmdDesc', 'aggrLevel', 'isLeaf', 'customsCode', 'customsDesc',
       'mosCode', 'mosDesc', 'motCode', 'motDesc', 'qtyUnitCode',
       'qtyUnitAbbr', 'qty', 'isQtyEstimated', 'altQtyUnitCode',
       'altQtyUnitAbbr', 'altQty', 'isAltQtyEstimated', 'netWgt',
       'isnetWgtEstimated', 'grossWgt', 'isgrossWgtEstimated', 'cifValue',
       'fobValue', 'primaryValue', 'legacyEstimationFlag', 'isReported',
       'isAggregate'], dtype=object)

# Codebooks

List of codebooks for the different columns

In [1]:
import comtradeapicall   # this is the UN supplied library

codebooks = comtradeapicall.listReference()
codebooks

category                variable  \
0             dataitem        Trade data items   
1                 freq               Frequency   
2                 flow              Trade Flow   
3                  mot      Mode of Transports   
4              partner    Partner country/area   
5             reporter                Reporter   
6              qtyunit       Units of quantity   
7                  mos          Mode of Supply   
8               cmd:B4                 Product   
9               cmd:B5                 Product   
10             customs  Customs Procedure Code   
11            cmd:EB02                 Product   
12            cmd:EB10                 Product   
13           cmd:EB10S                 Product   
14              cmd:EB                 Product   
15              cmd:H0                 Product   
16              cmd:H1                 Product   
17              cmd:H2                 Product   
18              cmd:H3                 Product   
19              cmd:H4                 Product   
20              cmd:H5                 Product   
21              cmd:H6                 Product   
22              cmd:HS                 Product   
23              cmd:S1                 Product   
24              cmd:S2                 Product   
25              cmd:S3                 Product   
26              cmd:S4                 Product   
27              cmd:SS                 Product   
28       ais:dataitems          AIS Data Items   
29  ais:countriesareas     AIS Countries/Areas   
30           ais:flows          AIS Trade Flow   
31     ais:vesseltypes        AIS Vessel Types   

                                          description  \
0    List of data items/variables used in UN Comtrade   
1       The time interval at which observations occur   
2   Trade flow or sub-flow (exports, re-exports, i...   
3   The mode of transport used when goods enter or...   
4   The partner country or geographic area for the...   
5   The country or geographic area to which the me...   
6   The unit for quantity and alternate quantity d...   
7   The mode of supply on delivery of services (tr...   
8   The classification of BEC Rev.4 – goods and se...   
9   The classification of BEC Rev.5 – goods or ser...   
10                   Customs or statistical procedure   
11       The classification of  EBOPS 2002 – services   
12        The classification of EBOPS 2010 – services   
13   The classification of EBOPS 2010 SDMX – services   
14    The classification of Combined EBOPS – services   
15              The classification of HS 1992 – goods   
16              The classification of HS 1996 – goods   
17              The classification of HS 2002 – goods   
18              The classification of HS 2007 – goods   
19              The classification of HS 2012 – goods   
20              The classification of HS 2017 – goods   
21              The classification of HS 2022 – goods   
22          The classification of Combined HS – goods   
23           The classification of SITC Rev.1 – goods   
24           The classification of SITC Rev.2 – goods   
25           The classification of SITC Rev.3 – goods   
26           The classification of SITC Rev.4 – goods   
27        The classification of Combined SITC – goods   
28  Data items used in experimental seaborne trade...   
29  Countries and Areas used in experimental seabo...   
30  Trade flows used in experimental seaborne trad...   
31  Vessel types used in experimental seaborne tra...   

                                              fileuri  
0   https://comtradeapi.un.org/files/v1/app/refere...  
1   https://comtradeapi.un.org/files/v1/app/refere...  
2   https://comtradeapi.un.org/files/v1/app/refere...  
3   https://comtradeapi.un.org/files/v1/app/refere...  
4   https://comtradeapi.un.org/files/v1/app/refere...  
5   https://comtradeapi.un.org/files/v1/app/refere...  
6   https://comtradeapi.un.org/files/v1/app/refere...  
7   https://comtradeapi.un.org/fi

Get specific codebook

In [22]:
cbk = 'cmd:H6'
codebook = comtradeapicall.getReference(cbk)
codebook.head(10)

id                                               text parent isLeaf  \
0   TOTAL                         Total - All H6 commodities      #      0   
1      01                                 01 - Animals; live  TOTAL      0   
2    0101      0101 - Horses, asses, mules and hinnies; live     01      0   
3  010121  010121 - Horses; live, pure-bred breeding animals   0101      1   
4  010129  010129 - Horses; live, other than pure-bred br...   0101      1   
5  010130                               010130 - Asses; live   0101      1   
6  010190                   010190 - Mules and hinnies; live   0101      1   
7    0102                        0102 - Bovine animals; live     01      0   
8  010221  010221 - Cattle; live, pure-bred breeding animals   0102      1   
9  010229  010229 - Cattle; live, other than pure-bred br...   0102      1   

   aggrlevel standardUnitAbbr  
0          0              n/a  
1          2              n/a  
2          4              n/a  
3          6                u  
4          6                u  
5          6                u  
6          6                u  
7          4              n/a  
8          6                u  
9          6                u

## List of partners (countries, regiões, blocks)

The list includes countries but also:
* trade blocks (EU, LAIA, Africa CAMEU ...)
* former countries (e.g. USSR, Yugoslavia, Czechoslovakia, East Germany, West Germany, Serbia and Montenegro, Sudan and South Sudan, etc.)
* combinations (USA+Porto Rico)

In [33]:
import comtradeapicall

partners = comtradeapicall.getReference('partner')
partners.head(10)

id                      text  PartnerCode               PartnerDesc  \
0    4               Afghanistan            4               Afghanistan   
1  472  Africa CAMEU region, nes          472  Africa CAMEU region, nes   
2  248            Åland Islands           248            Åland Islands    
3    8                   Albania            8                   Albania   
4   12                   Algeria           12                   Algeria   
5   16            American Samoa           16            American Samoa   
6   20                   Andorra           20                   Andorra   
7   24                    Angola           24                    Angola   
8  660                  Anguilla          660                  Anguilla   
9   10                Antarctica           10                Antarctica   

                                    partnerNote PartnerCodeIsoAlpha2  \
0                                   Afghanistan                   AF   
1  Africa CAMEU region, not elsewhere specified                  NaN   
2                                           NaN                   AX   
3                                       Albania                   AL   
4                                       Algeria                   DZ   
5                                American Samoa                   AS   
6                                       Andorra                   AD   
7                                        Angola                   AO   
8                                      Anguilla                   AI   
9                                    Antarctica                   AQ   

  PartnerCodeIsoAlpha3   entryEffectiveDate  isGroup entryExpiredDate  
0                  AFG  1900-01-01T00:00:00    False              NaN  
1                  _AC  1900-01-01T00:00:00    False              NaN  
2                  ALA  1900-01-01T00:00:00    False              NaN  
3                  ALB  1900-01-01T00:00:00    False              NaN  
4                  DZA  1900-01-01T00:00:00    False              NaN  
5                  ASM  1900-01-01T00:00:00    False              NaN  
6                  AND  1900-01-01T00:00:00    False              NaN  
7                  AGO  1900-01-01T00:00:00    False              NaN  
8                  AIA  1981-01-01T00:00:00    False              NaN  
9                  ATA  1900-01-01T00:00:00    False              NaN

List groups in partners list

In [25]:
partners[partners.isGroup].head(10)

id   text  PartnerCode PartnerDesc partnerNote PartnerCodeIsoAlpha2  \
304   0  World            0       World       World                  NaN   

    PartnerCodeIsoAlpha3   entryEffectiveDate  isGroup entryExpiredDate  
304                  W00  1901-01-01T00:00:00     True              NaN

## List of "reporters" 

Countries that provide information for the database

In [35]:
import comtradeapicall

reporters = comtradeapicall.getReference('reporter')
reporters.head(10)

id                          text  reporterCode  \
0    4                   Afghanistan             4   
1    8                       Albania             8   
2   12                       Algeria            12   
3   20                       Andorra            20   
4   24                        Angola            24   
5  660                      Anguilla           660   
6   28           Antigua and Barbuda            28   
7  886  Arab Rep. of Yemen (...1990)           886   
8   32                     Argentina            32   
9   51                       Armenia            51   

                   reporterDesc                   reporterNote  \
0                   Afghanistan                    Afghanistan   
1                       Albania                        Albania   
2                       Algeria                        Algeria   
3                       Andorra                        Andorra   
4                        Angola                         Angola   
5                      Anguilla                       Anguilla   
6           Antigua and Barbuda            Antigua and Barbuda   
7  Arab Rep. of Yemen (...1990)  Former Arab Republic of Yemen   
8                     Argentina                      Argentina   
9                       Armenia                        Armenia   

  reporterCodeIsoAlpha2 reporterCodeIsoAlpha3   entryEffectiveDate  isGroup  \
0                    AF                   AFG  1900-01-01T00:00:00    False   
1                    AL                   ALB  1900-01-01T00:00:00    False   
2                    DZ                   DZA  1900-01-01T00:00:00    False   
3                    AD                   AND  1900-01-01T00:00:00    False   
4                    AO                   AGO  1900-01-01T00:00:00    False   
5                    AI                   AIA  1981-01-01T00:00:00    False   
6                    AG                   ATG  1900-01-01T00:00:00    False   
7                    YE                   YEM  1900-01-01T00:00:00    False   
8                    AR                   ARG  1900-01-01T00:00:00    False   
9                    AM                   ARM  1992-01-01T00:00:00    False   

      entryExpiredDate  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
5                  NaN  
6                  NaN  
7  1990-12-31T00:00:00  
8                  NaN  
9                  NaN

In [36]:
reporters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id                     255 non-null    int64 
 1   text                   255 non-null    object
 2   reporterCode           255 non-null    int64 
 3   reporterDesc           255 non-null    object
 4   reporterNote           254 non-null    object
 5   reporterCodeIsoAlpha2  248 non-null    object
 6   reporterCodeIsoAlpha3  255 non-null    object
 7   entryEffectiveDate     255 non-null    object
 8   isGroup                255 non-null    bool  
 9   entryExpiredDate       33 non-null     object
dtypes: bool(1), int64(2), object(7)
memory usage: 18.3+ KB


In [37]:
# create a dict called country_codes with data from "partners" data frame
# and reporters data frame

country_codes =  dict(zip(reporters['reporterCode'], reporters['reporterDesc']))
country_codes.update(dict(zip(partners['PartnerCode'], partners['PartnerDesc'])))
country_codes


{4: 'Afghanistan',
 8: 'Albania',
 12: 'Algeria',
 20: 'Andorra',
 24: 'Angola',
 660: 'Anguilla',
 28: 'Antigua and Barbuda',
 886: 'Arab Rep. of Yemen (...1990)',
 32: 'Argentina',
 51: 'Armenia',
 533: 'Aruba',
 975: 'ASEAN',
 36: 'Australia',
 40: 'Austria',
 31: 'Azerbaijan',
 44: 'Bahamas',
 48: 'Bahrain',
 50: 'Bangladesh',
 52: 'Barbados',
 112: 'Belarus',
 56: 'Belgium',
 58: 'Belgium-Luxembourg (...1998)',
 84: 'Belize',
 204: 'Benin',
 60: 'Bermuda',
 64: 'Bhutan',
 68: 'Bolivia (Plurinational State of)',
 535: 'Bonaire',
 70: 'Bosnia Herzegovina',
 72: 'Botswana',
 92: 'Br. Virgin Isds',
 76: 'Brazil',
 96: 'Brunei Darussalam',
 100: 'Bulgaria',
 854: 'Burkina Faso',
 108: 'Burundi',
 132: 'Cabo Verde',
 116: 'Cambodia',
 120: 'Cameroon',
 124: 'Canada',
 136: 'Cayman Isds',
 140: 'Central African Rep.',
 148: 'Chad',
 152: 'Chile',
 156: 'China',
 344: 'China, Hong Kong SAR',
 446: 'China, Macao SAR',
 170: 'Colombia',
 174: 'Comoros',
 178: 'Congo',
 184: 'Cook Isds',
 18

##  Modes of transport

In [38]:
import comtradeapicall

comtradeapicall.getReference('mot')


id                                           text
0      0                       TOTAL modes of transport
1   1000                                            Air
2   2000                                          Water
3   2100                                            Sea
4   2200                                Inland waterway
5   2900                     Water, not else classified
6   3000                                           Land
7   3100                                        Railway
8   3200                                           Road
9   3900                      Land, not else classified
10  9000                       Not elsewhere classified
11  9100                           Pipelines and cables
12  9110                                      Pipelines
13  9120                                         Cables
14  9190      Pipelines and cables, not else classified
15  9200  Postal consignments, mail or courier shipment
16  9300                           Self propelled goods
17  9900                                          Other

## Flow

In [39]:
import comtradeapicall

comtradeapicall.getReference('flow')

id                                      text
0    M                                    Import
1    X                                    Export
2   DX                           Domestic Export
3   FM                            Foreign Import
4  MIP     Import of goods for inward processing
5  MOP  Import of goods after outward processing
6   RM                                 Re-import
7   RX                                 Re-export
8  XIP   Export of goods after inward processing
9  XOP    Export of goods for outward processing

## HS Codes

In [43]:
import comtradeapicall


hs_codes = comtradeapicall.getReference('cmd:H5') # also cmd:H2, cmd:H6


Filter HS codes

In [50]:
import pandas as pd
pd.set_option('display.max_colwidth', 200)

text = 'telephone'

hs_codes[hs_codes['text'].str.contains(text, case=False)]

id  \
5867    8517   
5868  851711   
5869  851712   
5870  851718   
5872  851762   
5873  851769   
5874  851770   
5887  851950   

                                                                                                                                                                                                         text  \
5867  8517 - Telephone sets, including telephones for cellular networks or for other wireless networks; other apparatus for the transmission or reception of voice, images or other data (including wired/...   
5868                                                                                                                                                      851711 - Line telephone sets with cordless handsets   
5869                                                                                                                                 851712 - Telephones for cellular networks or for other wireless networks   
5870                                                                                                                                                        851718 - Telephone sets n.e.c. in item no. 8517.1   
5872  851762 - Communication apparatus (excluding telephone sets or base stations); machines for the reception, conversion and transmission or regeneration of voice, images or other data, including swit...   
5873  851769 - Communication apparatus (excluding telephone sets or base stations); machines for the transmission or reception of voice, images or other data (including wired/wireless networks), n.e.c. ...   
5874                                                     851770 - Telephone sets and other apparatus for the transmission or reception of voice, images or other data, via a wired or wireless network; parts   
5887                                                                                                                          851950 - Sound recording or reproducing apparatus; telephone answering machines   

     parent isLeaf  aggrlevel standardUnitAbbr  
5867     85      0          4              n/a  
5868   8517      1          6                u  
5869   8517      1          6                u  
5870   8517      1          6                u  
5872   8517      1          6                u  
5873   8517      1          6                u  
5874   8517      1          6               kg  
5887   8519      1          6                u

# Tests

Reproducing calls to python API from comtrade. 

comtradetools.getFinalData() interfaces to UN supplied comtradeapicall.getFinalData() adding:
* rate management, so that too many calls errors are avoided
* long lists of periods: original function can only take 12 periods, comtradetools breaks longer lists into less
   than 12 and concatenates the results
* cache management: calls are cached to avoid repetitive access to Comtrade servers.

See: https://github.com/uncomtrade/comtradeapicall

* Extract Australia imports of commodity codes 90 and 91 from all partners in classic mode in May 2022

In [52]:
import comtradeapicall
import comtradetools

APIKEY = comtradetools.get_api_key()

mydf = comtradeapicall.getFinalData(APIKEY, typeCode='C', freqCode='M', clCode='HS', period='202205',
                                    reporterCode='36', cmdCode='91,90', flowCode='M', partnerCode=None,
                                    partner2Code=None,
                                    customsCode=None, motCode=None, maxRecords=2500, format_output='JSON',
                                    aggregateBy=None, breakdownMode='classic', countOnly=None, includeDesc=True)
mydf.head()

typeCode freqCode  refPeriodId  refYear  refMonth  period  reporterCode  \
0        C        M     20220501     2022         5  202205            36   
1        C        M     20220501     2022         5  202205            36   
2        C        M     20220501     2022         5  202205            36   
3        C        M     20220501     2022         5  202205            36   
4        C        M     20220501     2022         5  202205            36   

  reporterISO reporterDesc flowCode  ... netWgt  isNetWgtEstimated grossWgt  \
0         AUS    Australia        M  ...    0.0              False      0.0   
1         AUS    Australia        M  ...    0.0              False      0.0   
2         AUS    Australia        M  ...    0.0               True      0.0   
3         AUS    Australia        M  ...    0.0               True      0.0   
4         AUS    Australia        M  ...    0.0               True      0.0   

  isGrossWgtEstimated      cifvalue      fobvalue  primaryValue  \
0               False  8.388227e+08  8.012504e+08  8.388227e+08   
1               False  1.852201e+03  1.726534e+03  1.852201e+03   
2               False  3.695805e+03  3.224809e+03  3.695805e+03   
3               False  1.190824e+05  1.114879e+05  1.190824e+05   
4               False  6.946143e+06  6.808902e+06  6.946143e+06   

  legacyEstimationFlag isReported  isAggregate  
0                    0      False         True  
1                    0      False         True  
2                    4      False         True  
3                    4      False         True  
4                    4      False         True  

[5 rows x 47 columns]

In [12]:
import comtradetools

mydf = comtradetools.getFinalData(APIKEY,
                                    typeCode='C',
                                    freqCode='M',
                                    clCode='HS',
                                    period='202205',
                                    reporterCode='36',
                                    cmdCode='91,90',
                                    flowCode='M',
                                    partnerCode=None,
                                    partner2Code=None,
                                    customsCode=None,
                                    motCode=None,
                                    maxRecords=2500,
                                    format_output='JSON',
                                    aggregateBy=None,
                                    breakdownMode='classic',
                                    countOnly=None,
                                    includeDesc=True)
mydf.head()

INFO:root:Calling getFinalData for period 202205


typeCode freqCode  refPeriodId  refYear  refMonth  period  reporterCode  \
0        C        M     20220501     2022         5  202205            36   
1        C        M     20220501     2022         5  202205            36   
2        C        M     20220501     2022         5  202205            36   
3        C        M     20220501     2022         5  202205            36   
4        C        M     20220501     2022         5  202205            36   

  reporterISO reporterDesc flowCode  ... netWgt  isNetWgtEstimated grossWgt  \
0         AUS    Australia        M  ...    0.0              False      0.0   
1         AUS    Australia        M  ...    0.0              False      0.0   
2         AUS    Australia        M  ...    0.0               True      0.0   
3         AUS    Australia        M  ...    0.0               True      0.0   
4         AUS    Australia        M  ...    0.0               True      0.0   

  isGrossWgtEstimated      cifvalue      fobvalue  primaryValue  \
0               False  8.388227e+08  8.012504e+08  8.388227e+08   
1               False  1.852201e+03  1.726534e+03  1.852201e+03   
2               False  3.695805e+03  3.224809e+03  3.695805e+03   
3               False  1.190824e+05  1.114879e+05  1.190824e+05   
4               False  6.946143e+06  6.808902e+06  6.946143e+06   

  legacyEstimationFlag isReported  isAggregate  
0                    0      False         True  
1                    0      False         True  
2                    4      False         True  
3                    4      False         True  
4                    4      False         True  

[5 rows x 47 columns]

## Países de consignação

### Partner2

O `partner2` é uma novidade dos dados comtrade. Procura registar o país de "consignação".

No caso das importações é o país que despachou os bens para o país que importa,
sem que tenha ocorrido entre o país de origem (_partner_) nenhuma transação
que modifique o estatuto legal (denominação de origem?) dos bens:

> The country of consignment in the case of imports is the country from which goods
 were dispatched to the importing country, without any commercial transactions 
 or other operations that change the legal status of the goods taking 
 place in any intermediate country.


#### Resultados quand o partner2Code não é especificado na pesquisa



Se a pesquisa não especifica partner2Code, alguns anos produzem mais do que uma linha por par _reporter/partner_  
com diferentes valores. Por exemplo, se a China for o `reporter` e a Guiné Equatorial o `partner` nos anos 2015, 2016, 2017 aparece::
* Uma linha por `partner2Code`, incluindo uma linha em que o `partner2` é igual ao `partner` (importações diretas).
* Uma linha adicional com `partner2Code` igual a zero que contém o total agregado das outras linhas com `partner2Code` explícito.
* Isso significa que existe duplicação do total.
  
|    | reporterDesc   | partnerDesc       |   partner2Code | partner2Desc         |   refYear | cmdCode   | flowCode   | primaryValueFormated   |
|---:|:---------------|:------------------|---------------:|:---------------------|----------:|:----------|:-----------|:-----------------------|
|  3 | China          | Equatorial Guinea |            344 | China, Hong Kong SAR |      2015 | TOTAL     | M          | 59.0                   |
|  1 | China          | Equatorial Guinea |             56 | Belgium              |      2015 | TOTAL     | M          | 2,435.0                |
|  2 | China          | Equatorial Guinea |            226 | Equatorial Guinea    |      2015 | TOTAL     | M          | 1,166,493,970.0        |
|  0 | China          | Equatorial Guinea |              0 | nan                  |      2015 | TOTAL     | M          | 1,166,496,464.0        |


Para evitar isso tem de se chamar a API com partner2Code = 0, para que os resultados de 2015,2016,2017 excluam
a decomposição. Se partner2Code=None as linhas adicionais aparecem.


Exemplo de resultados se o `partner2Code` for None.

Alterar as variáveis seguintes para testar.


In [18]:
import comtradetools

period = "2015" ## if freqCode M  use aaaamm
flow = "M"
cmdCode = 'TOTAL'
reporterCode = comtradetools.m49_china
partnerCode = comtradetools.m49_guine_equatorial
# None: total and subtotal per country
# 0: only total
# > 0: only this country code
partner2Code = None

Exemplo como com `partner2Code = None` aparecem linhas
com detalhe de partner linhas de total

In [19]:
import comtradetools
import pandas as pd


APIKEY = comtradetools.get_api_key()

partner2Code = None

pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 500)
pd.options.display.float_format = '{:,.2f}'.format

cols_partner2 = ['reporterDesc','partnerDesc','partner2Code','partner2Desc','refYear',
        'flowDesc','primaryValue','cmdCode','qty','qtyUnitCode','qtyUnitAbbr','customsCode','motDesc']
cols_partner2_qty_weight = ['reporterDesc','partnerDesc','partner2Code','partner2Desc','refYear','cmdDesc',
        'flowCode','primaryValue',
        'qty','qtyUnitCode','qty','isQtyEstimated','altQtyUnitCode', 'altQtyUnitAbbr', 'altQty', 'isAltQtyEstimated',
       'netWgt', 'isNetWgtEstimated', 'grossWgt', 'isGrossWgtEstimated',
        'motCode']

df = comtradetools.getFinalData(
                APIKEY,
                typeCode="C",# C for commodities, S for Services
                freqCode="A",# (freqCode) A for annual and M for monthly
                flowCode=flow,
                reporterCode=reporterCode,
                partnerCode=partnerCode,
                partner2Code=partner2Code,
                clCode='HS',
                cmdCode=cmdCode,
                customsCode=None,
                motCode=None,
                period=period,
                includeDesc=True,

                )
result = df.sort_values(['partner2Code','primaryValue'], ascending=[True,False])[cols_partner2]
# print(result.to_markdown())
result

INFO:root:Calling getFinalData for period 2015
INFO:root:Using cached results for period 2015


reporterDesc        partnerDesc  partner2Code          partner2Desc  \
0        China  Equatorial Guinea             0                 World   
1        China  Equatorial Guinea            56               Belgium   
2        China  Equatorial Guinea           226     Equatorial Guinea   
3        China  Equatorial Guinea           344  China, Hong Kong SAR   

   refYear flowDesc     primaryValue cmdCode  qty  qtyUnitCode qtyUnitAbbr  \
0     2015   Import 1,166,496,464.00   TOTAL 0.00           -1         N/A   
1     2015   Import         2,435.00   TOTAL 0.00           -1         N/A   
2     2015   Import 1,166,493,970.00   TOTAL 0.00           -1         N/A   
3     2015   Import            59.00   TOTAL 0.00           -1         N/A   

  customsCode    motDesc  
0         C00  TOTAL MOT  
1         C00  TOTAL MOT  
2         C00  TOTAL MOT  
3         C00  TOTAL MOT

A função `comtradetools.getFinalData` passou a colocar `partner2Code = 0` quando não especificado,
para evitar o problema.

Neste exemplo não se inclui o parâmetro `partner2Code` e a função coloca a zero para obter o resultado
correcto.

In [20]:
import comtradetools

df = comtradetools.getFinalData(
                APIKEY,
                typeCode="C",# C for commodities, S for Services
                freqCode="A",# (freqCode) A for annual and M for monthly
                flowCode=flow,
                reporterCode=comtradetools.m49_china,
                partnerCode=partnerCode,
                motCode=None,
                customsCode=None,
                clCode='HS',
                cmdCode='TOTAL',
                period=period,
                includeDesc=True
                )
result = df.sort_values(['partnerDesc','flowCode'])[cols_partner2]
# print(result.to_markdown())
result

INFO:root:Calling getFinalData for period 2015
INFO:root:Using cached results for period 2015


reporterDesc        partnerDesc  partner2Code partner2Desc  refYear  \
0        China  Equatorial Guinea             0        World     2015   

  flowDesc     primaryValue cmdCode  qty  qtyUnitCode qtyUnitAbbr customsCode  \
0   Import 1,166,496,464.00   TOTAL 0.00           -1         N/A         C00   

     motDesc  
0  TOTAL MOT

#### Cobertura de informação sobre parter2

Não parece ser possível obter as informações referentes a partner2
 senão nos anos 2015-2017, quando o `reporter` é a China.

In [23]:
import comtradetools

yrange = comtradetools.year_range(2014,2023)
reporterCode = comtradetools.m49_china
partnerCode = comtradetools.m49_guine_equatorial


In [24]:

df = comtradetools.getFinalData(
    typeCode="C",# C for commodities, S for Services
    freqCode="A",# (freqCode) A for annual and M for monthly
    flowCode=flow,
    reporterCode=reporterCode,
    partnerCode=partnerCode,
    partner2Code=None,
    motCode=None,
    customsCode=None,
    clCode='HS',
    cmdCode='TOTAL',
    period=yrange,
    includeDesc=True,
    )
result = df.sort_values(['partnerDesc','refYear','flowCode','motDesc'])[cols_partner2]
# print(result.to_markdown())
result

INFO:root:Calling getFinalData for period 2014,2015,2016,2017,2018,2019,2020,2021,2022,2023
INFO:root:Using cached results for period 2014,2015,2016,2017,2018,2019,2020,2021,2022,2023


reporterDesc        partnerDesc  partner2Code          partner2Desc  \
0         China  Equatorial Guinea             0                 World   
1         China  Equatorial Guinea             0                 World   
2         China  Equatorial Guinea            56               Belgium   
3         China  Equatorial Guinea           226     Equatorial Guinea   
4         China  Equatorial Guinea           344  China, Hong Kong SAR   
5         China  Equatorial Guinea             0                 World   
6         China  Equatorial Guinea            24                Angola   
7         China  Equatorial Guinea           178                 Congo   
8         China  Equatorial Guinea           226     Equatorial Guinea   
9         China  Equatorial Guinea           251                France   
10        China  Equatorial Guinea           344  China, Hong Kong SAR   
11        China  Equatorial Guinea           458              Malaysia   
12        China  Equatorial Guinea           490       Other Asia, nes   
13        China  Equatorial Guinea             0                 World   
14        China  Equatorial Guinea            24                Angola   
15        China  Equatorial Guinea           124                Canada   
16        China  Equatorial Guinea           226     Equatorial Guinea   
17        China  Equatorial Guinea           344  China, Hong Kong SAR   
18        China  Equatorial Guinea           392                 Japan   
19        China  Equatorial Guinea           458              Malaysia   
20        China  Equatorial Guinea           704              Viet Nam   
21        China  Equatorial Guinea             0                 World   
22        China  Equatorial Guinea             0                 World   
23        China  Equatorial Guinea             0                 World   
24        China  Equatorial Guinea             0                 World   
25        China  Equatorial Guinea             0                 World   
26        China  Equatorial Guinea             0                 World   

    refYear flowDesc     primaryValue cmdCode  qty  qtyUnitCode qtyUnitAbbr  \
0      2014   Import 3,217,190,248.00   TOTAL  NaN           -1         N/A   
1      2015   Import 1,166,496,464.00   TOTAL 0.00           -1         N/A   
2      2015   Import         2,435.00   TOTAL 0.00           -1         N/A   
3      2015   Import 1,166,493,970.00   TOTAL 0.00           -1         N/A   
4      2015   Import            59.00   TOTAL 0.00           -1         N/A   
5      2016   Import   631,851,506.00   TOTAL 0.00           -1         N/A   
6      2016   Import       396,344.00   TOTAL 0.00           -1         N/A   
7      2016   Import     1,457,849.00   TOTAL 0.00           -1         N/A   
8      2016   Import   589,959,003.00   TOTAL 0.00           -1         N/A   
9      2016   Import         1,341.00   TOTAL 0.00           -1         N/A   
10     2016   Import           634.00   TOTAL 0.00           -1         N/A   
11     2016   Import    40,036,318.00   TOTAL 0.00           -1         N/A   
12     2016   Import            17.00   TOTAL 0.00           -1         N/A   
13     2017   Import 1,483,835,026.00   TOTAL 0.00           -1         N/A   
14     2017   Import         8,468.00   TOTAL 0.00           -1         N/A   
15     2017   Import            20.00   TOTAL 0.00           -1         N/A   
16     2017   Import 1,422,618,641.00   TOTAL 0.00           -1         N/A   
17     2017   Import            46.00   TOTAL 0.00           -1         N/A   
18     2017   Import            10.00   TOTAL 0.00           -1         N/A   
19     2017   Import    59,992,784.00   TOTAL 0.00           -1         N/A   
20     2017   Import     1,215,057.00   TOTAL 0.00           -1         N/A   
21     2018   Import 2,139,372,096.00   TOTAL 0.00           -1         N/A   
22     2019   Import 1,723,064,057.00   TOTAL 0.00           -1         N/A   
23     2020   Import 1,182,549,225

Mas outros países incluem esses dados em outros anos, por exemplo, Portugal.

No exemplo com reporter = Portugal e partner = Brasil não só se obtém linhas a discriminar
o partner2 mas também, sem motCode=0, linhas a discriminar os "Mode of Transport"

In [26]:
import comtradetools

period = "2018" ## if freqCode M  use aaaamm
flow = "X"
reporterCode=comtradetools.m49_portugal
partnerCode = comtradetools.m49_brazil
cmdCode='TOTAL'

df = comtradetools.getFinalData(
                    typeCode="C",# C for commodities, S for Services
                    freqCode="A",# (freqCode) A for annual and M for monthly
                    flowCode=flow,
                    reporterCode=reporterCode,
                    partnerCode=partnerCode,
                    partner2Code=None,
                    motCode=None,
                    customsCode=None,
                    clCode='HS',
                    cmdCode=cmdCode,
                    period=period,
                    includeDesc=True,
                    )
interesting_cols = [col for col in df.columns if len(df[col].unique())>1]
print(interesting_cols)
show_cols = cols_partner2 + list(set(interesting_cols)-set(cols_partner2))
result = df.sort_values(['partner2Code','flowCode'])[show_cols]

# print(result.to_markdown())
result

INFO:root:Calling getFinalData for period 2018
INFO:root:Using cached results for period 2018


['partner2Code', 'partner2ISO', 'partner2Desc', 'motCode', 'motDesc', 'isNetWgtEstimated', 'fobvalue', 'primaryValue']


reporterDesc partnerDesc  partner2Code    partner2Desc  refYear flowDesc  \
0      Portugal      Brazil             0           World     2018   Export   
1      Portugal      Brazil             0           World     2018   Export   
2      Portugal      Brazil             0           World     2018   Export   
3      Portugal      Brazil             0           World     2018   Export   
4      Portugal      Brazil             0           World     2018   Export   
5      Portugal      Brazil           251          France     2018   Export   
6      Portugal      Brazil           251          France     2018   Export   
7      Portugal      Brazil           276         Germany     2018   Export   
8      Portugal      Brazil           276         Germany     2018   Export   
9      Portugal      Brazil           276         Germany     2018   Export   
10     Portugal      Brazil           300          Greece     2018   Export   
11     Portugal      Brazil           300          Greece     2018   Export   
12     Portugal      Brazil           380           Italy     2018   Export   
13     Portugal      Brazil           380           Italy     2018   Export   
14     Portugal      Brazil           528     Netherlands     2018   Export   
15     Portugal      Brazil           528     Netherlands     2018   Export   
16     Portugal      Brazil           642         Romania     2018   Export   
17     Portugal      Brazil           642         Romania     2018   Export   
18     Portugal      Brazil           724           Spain     2018   Export   
19     Portugal      Brazil           724           Spain     2018   Export   
20     Portugal      Brazil           724           Spain     2018   Export   
21     Portugal      Brazil           826  United Kingdom     2018   Export   
22     Portugal      Brazil           826  United Kingdom     2018   Export   
23     Portugal      Brazil           899      Areas, nes     2018   Export   
24     Portugal      Brazil           899      Areas, nes     2018   Export   
25     Portugal      Brazil           899      Areas, nes     2018   Export   
26     Portugal      Brazil           899      Areas, nes     2018   Export   
27     Portugal      Brazil           899      Areas, nes     2018   Export   

       primaryValue cmdCode  qty  qtyUnitCode qtyUnitAbbr customsCode  \
0  1,083,984,874.36   TOTAL 0.00           -1         N/A         C00   
1     52,669,113.05   TOTAL 0.00           -1         N/A         C00   
2  1,027,354,167.57   TOTAL 0.00           -1         N/A         C00   
3      3,801,817.75   TOTAL 0.00           -1         N/A         C00   
4        159,776.00   TOTAL 0.00           -1         N/A         C00   
5         56,538.91   TOTAL 0.00           -1         N/A         C00   
6         56,538.91   TOTAL 0.00           -1         N/A         C00   
7         18,143.93   TOTAL 0.00           -1         N/A         C00   
8         16,982.24   TOTAL 0.00           -1         N/A         C00   
9          1,161.69   TOTAL 0.00           -1         N/A         C00   
10       235,517.71   TOTAL 0.00           -1         N/A         C00   
11       235,517.71   TOTAL 0.00           -1         N/A         C00   
12        50,945.52   TOTAL 0.00           -1         N/A         C00   
13        50,945.52   TOTAL 0.00           -1         N/A         C00   
14        15,883.18   TOTAL 0.00           -1         N/A         C00   
15        15,883.18   TOTAL 0.00           -1         N/A         C00   
16        26,405.78   TOTAL 0.00           -1         N/A         C00   
17        26,405.78   TOTAL 0.00           -1         N/A         C00   
18    16,145,881.59   TOTAL 0.00           -1         N/A         C00   
19     3,550,776.12   TOTAL 0.00           -1         N/A         C00   
20    12,595,105.47   TOTAL 0.00           -1         N/A         C00   
21         1,626.13   TOTAL 0.00           -1         N/A         C00   
22         1,626.13   TOTAL 0.00

Note-se a duplicação de linhas para o mesmo terceto _reporterCode-partnerCode-partner2Code_

Essa duplicação deve-se a desdobramento `motCode` (_mode of transport_) e `partner2Code`. Se filtrarmos só por motCode == 0 
e partner2Code != 0 obtemos a lista de combinações reporter-partner-partner2 para todos os modos de tranporte, sem
valores agregados por partner.

In [27]:
result[(result.motCode==0)&(result.partner2Code != 0)][list(set(['refYear','reporterDesc','partnerDesc',]+interesting_cols))]

partnerDesc  refYear    partner2Desc     primaryValue partner2ISO  \
5       Brazil     2018          France        56,538.91         FRA   
7       Brazil     2018         Germany        18,143.93         DEU   
10      Brazil     2018          Greece       235,517.71         GRC   
12      Brazil     2018           Italy        50,945.52         ITA   
14      Brazil     2018     Netherlands        15,883.18         NLD   
16      Brazil     2018         Romania        26,405.78         ROU   
18      Brazil     2018           Spain    16,145,881.59         ESP   
21      Brazil     2018  United Kingdom         1,626.13         GBR   
23      Brazil     2018      Areas, nes 1,067,433,931.61         _X    

   reporterDesc  partner2Code  motCode    motDesc  isNetWgtEstimated  \
5      Portugal           251        0  TOTAL MOT              False   
7      Portugal           276        0  TOTAL MOT              False   
10     Portugal           300        0  TOTAL MOT              False   
12     Portugal           380        0  TOTAL MOT              False   
14     Portugal           528        0  TOTAL MOT              False   
16     Portugal           642        0  TOTAL MOT              False   
18     Portugal           724        0  TOTAL MOT               True   
21     Portugal           826        0  TOTAL MOT              False   
23     Portugal           899        0  TOTAL MOT               True   

           fobvalue  
5         56,538.91  
7         18,143.93  
10       235,517.71  
12        50,945.52  
14        15,883.18  
16        26,405.78  
18    16,145,881.59  
21         1,626.13  
23 1,067,433,931.61

Gravar em Excel (nome do ficheiro automaticamente reflecte o valor das variáveis relevantes)

In [31]:
import comtradetools

result.to_excel(f"./reports/partner2_{comtradetools.decode_country(reporterCode)}_{comtradetools.decode_country(partnerCode)}_{flow}_{period}.xlsx")

## Testes



In [34]:
import comtradetools

period = "2016" ## if freqCode M  use aaaamm
flow = "M,X"
cmdCode = 'TOTAL'
reporterCode = comtradetools.m49_portugal
partnerCode = comtradetools.m49_angola

pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 500)
pd.options.display.float_format = '${:,.2f}'.format

cols_partner2 = ['reporterDesc','partnerDesc','partner2Code','partner2Desc','refYear',
        'flowCode','qtyUnitCode','primaryValue','isAggregate']
cols_partner2_qty_weight = ['reporterDesc','partnerDesc','partner2Code','partner2Desc','refYear','cmdDesc',
        'flowCode','primaryValue',
        'qtyUnitCode','qty','isQtyEstimated','altQtyUnitCode', 'altQtyUnitAbbr', 'altQty', 'isAltQtyEstimated',
       'netWgt', 'isNetWgtEstimated', 'grossWgt', 'isGrossWgtEstimated',
        'motCode']


df = comtradetools.getFinalData(
        typeCode="C",# C for commodities, S for Services
        freqCode="A",# (freqCode) A for annual and M for monthly
        flowCode=flow,
        period=period,
        reporterCode=reporterCode,
        partnerCode=partnerCode,
        partner2Code=None,
        motCode=None,
        customsCode=None,
        clCode='HS',
        cmdCode=cmdCode,
        includeDesc=True
        )

interesting_cols = [col for col in df.columns if len(df[col].unique())>1]
print(interesting_cols)
show_cols = cols_partner2 +  list(set(interesting_cols)-set(cols_partner2))
result = df.sort_values(['partnerDesc','flowCode'])[show_cols]
result

INFO:root:Calling getFinalData for period 2016


['flowCode', 'flowDesc', 'partner2Code', 'partner2ISO', 'partner2Desc', 'motCode', 'motDesc', 'isNetWgtEstimated', 'cifvalue', 'fobvalue', 'primaryValue']


reporterDesc partnerDesc  partner2Code          partner2Desc  refYear  \
0      Portugal      Angola             0                 World     2016   
2      Portugal      Angola             0                 World     2016   
4      Portugal      Angola             0                 World     2016   
8      Portugal      Angola             0                 World     2016   
10     Portugal      Angola             0                 World     2016   
12     Portugal      Angola            24                Angola     2016   
13     Portugal      Angola            24                Angola     2016   
14     Portugal      Angola            24                Angola     2016   
15     Portugal      Angola            24                Angola     2016   
16     Portugal      Angola            24                Angola     2016   
31     Portugal      Angola           446      China, Macao SAR     2016   
32     Portugal      Angola           446      China, Macao SAR     2016   
36     Portugal      Angola           516               Namibia     2016   
37     Portugal      Angola           516               Namibia     2016   
51     Portugal      Angola           784  United Arab Emirates     2016   
52     Portugal      Angola           784  United Arab Emirates     2016   
1      Portugal      Angola             0                 World     2016   
3      Portugal      Angola             0                 World     2016   
5      Portugal      Angola             0                 World     2016   
6      Portugal      Angola             0                 World     2016   
7      Portugal      Angola             0                 World     2016   
9      Portugal      Angola             0                 World     2016   
11     Portugal      Angola             0                 World     2016   
17     Portugal      Angola            40               Austria     2016   
18     Portugal      Angola            40               Austria     2016   
19     Portugal      Angola            56               Belgium     2016   
20     Portugal      Angola            56               Belgium     2016   
21     Portugal      Angola            56               Belgium     2016   
22     Portugal      Angola           208               Denmark     2016   
23     Portugal      Angola           208               Denmark     2016   
24     Portugal      Angola           251                France     2016   
25     Portugal      Angola           251                France     2016   
26     Portugal      Angola           276               Germany     2016   
27     Portugal      Angola           276               Germany     2016   
28     Portugal      Angola           380                 Italy     2016   
29     Portugal      Angola           380                 Italy     2016   
30     Portugal      Angola           380                 Italy     2016   
33     Portugal      Angola           470                 Malta     2016   
34     Portugal      Angola           470                 Malta     2016   
35     Portugal      Angola           470                 Malta     2016   
38     Portugal      Angola           528           Netherlands     2016   
39     Portugal      Angola           528           Netherlands     2016   
40     Portugal      Angola           528           Netherlands     2016   
41     Portugal      Angola           620              Portugal     2016   
42     Portugal      Angola           620              Portugal     2016   
43     Portugal      Angola           620              Portugal     2016   
44     Portugal      Angola           620              Portugal     2016   
45     Portugal      Angola           620              Portugal     2016   
46     Portugal      Angola           620              Portugal     2016   
47     Portugal      Angola           620              Portugal     2016   
48     Portugal      Angola           724                 Spain     2016   
49     Portugal      Angola           724          